src/commissioner/Discoverer.cpp

```diff
@@ -0,0 +1,264 @@
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR DnssdDiscoverer::GetAddress(chip::Optional<Transport::PeerAddress> & address,
+                                       chip::Optional<ReliableMessageProtocolConfig> & mrpConfig)
+{
+    if (mNodeIdx >= GetDiscoveredNodes().size() || mIpIdx >= GetDiscoveredNodes().data()[mNodeIdx].numIPs)
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    Dnssd::DiscoveredNodeData & node = GetDiscoveredNodes().data()[mNodeIdx];
+    Inet::InterfaceId interfaceId = node.ipAddress[mIpIdx].IsIPv6LinkLocal() ? node.interfaceId[mIpIdx] : Inet::InterfaceId::Null();
+#if 0
+        // TODO(#13326): connecting via TCP does not work
+        if (node.supportsTcp)
+        {
+            address.SetValue(Transport::PeerAddress::TCP(node.ipAddress[mIpIdx], node.port, interfaceId));
+            mrpConfig.SetValue(node.GetMRPConfig());
+        }
+        else
+#endif
+    {
+        address.SetValue(Transport::PeerAddress::UDP(node.ipAddress[mIpIdx], node.port, interfaceId));
+        mrpConfig.SetValue(node.GetMRPConfig());
+    }
+    return CHIP_NO_ERROR;
+}
+
+#endif // CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+
+Discoverer::Discoverer(Controller::DeviceControllerSystemState & systemState, Platform::SharedPtr<SetupPayload> payload,
+                       Delegate * delegate) :
+    Joinable(&mDelegate),
+    mDelegate(delegate), mPayload(payload)
+#if CONFIG_NETWORK_LAYER_BLE
+    ,
+    mBleDiscoverer(systemState, &mDelegate)
+#endif
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+    ,
+    mDnssdDiscoverer(systemState, &mDelegate)
+#endif
+{}
+
+CHIP_ERROR Discoverer::Init()
+{
+    RetainShutdownToken(Platform::MakeShared<ShutdownToken>(&mDelegate));
+    return InProgress() ? CHIP_NO_ERROR : CHIP_ERROR_NO_MEMORY;
+}
+
+void Discoverer::Shutdown()
+{
+#if CONFIG_NETWORK_LAYER_BLE
+    mBleDiscoverer.Shutdown();
+#endif
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+    mDnssdDiscoverer.Shutdown();
+#endif
+    ReleaseShutdownToken();
+}
+
+void Discoverer::SetDelegate(Delegate * delegate)
+{
+    mDelegate = delegate;
+}
+
+CHIP_ERROR Discoverer::Discover()
+{
+    CHIP_ERROR err = CHIP_ERROR_NOT_IMPLEMENTED;
+#if CONFIG_NETWORK_LAYER_BLE
+    bool searchAllOver = this->mPayload.get()->rendezvousInformation == RendezvousInformationFlag::kNone;
+    if (searchAllOver || this->mPayload.get()->rendezvousInformation == RendezvousInformationFlag::kBLE)
+    {
+        SuccessOrExit(err = mBleDiscoverer.StartBleDiscovery(*mPayload.get(), GetShutdownToken()));
+    }
+#endif
+#if CHIP_DEVICE_CONFIG_ENABLE_DNSSD
+    // We always want to search on-network because any node that has
+    // already been commissioned will use on-network regardless of
+    // onboarding payload contents.
+    SuccessOrExit(err = mDnssdDiscoverer.StartDnssdDiscovery(*mPayload.get(), GetShutdownToken()));
+#endif
+exit:
+    return err;
+}
+
+CHIP_ERROR Discoverer::GetNextCandidate(Commissionee & commissionee)
```

Comment:
In the case where both BLE and DNSSD are supported, what is the expected behavior here for this function?

Reply:
This returns any available BLE candidate first, and then the next available DNS-SD candidate if any is available.

This very purposefully has side effects.  It makes it such that the Discoverer keeps track of already-consumed candidates vs. next candidates so that the caller only sees each candidate once.

This removes the need for the caller to have to keep track of any of that.  Will add docs.

Reply:
Comment headers added.